src/store/s3store.rs

```diff
@@ -0,0 +1,193 @@
+        Some(_) => {
+            let access_key = url.username().to_string();
+            let access_secret = url
+                .password()
+                .ok_or(Error::InvalidConfigs(String::from(
+                    "did not find secret key",
+                )))?
+                .to_string();
+            let host = url
+                .host_str()
+                .ok_or(Error::InvalidConfigs(String::from("did not find host")))?
+                .to_string();
+            let port = url
+                .port()
+                .ok_or(Error::InvalidConfigs(String::from("did not find port")))?;
+
+            // rust-s3 implementation force tls unless it found `://` in the endpoint
+            // check rust-s3/aws-region/src/region.rs #fn scheme
+            let scheme = match url.query_pairs().find(|(key, _)| key == "tls") {
+                Some((_, value)) if value == "false" => "http://",
+                _ => "",
+            };
+
+            let endpoint = format!("{}{}:{}", scheme, host, port);
+
+            let bucket_name = url.path().trim_start_matches('/').to_string();
+
+            let region = match url.query_pairs().find(|(key, _)| key == "region") {
+                Some((_, value)) => value.to_string(),
+                None => return Err(Error::InvalidConfigs(String::from("did not find region"))),
+            };
+
+            (access_key, access_secret, endpoint, bucket_name, region)
+        }
+        None => {
+            return Err(Error::InvalidConfigs(String::from(
+                "failed parsing the url",
+            )))
+        }
+    };
+
+    Ok((
+        Credentials {
+            access_key: Some(access_key),
+            secret_key: Some(access_secret),
+            security_token: None,
+            session_token: None,
+            expiration: None,
+        },
+        Region::Custom {
+            region: region_name,
+            endpoint: endpoint,
+        },
+        bucket_name,
+    ))
+}
+
+async fn make_inner(url: String) -> Result<Box<dyn Store>> {
+    let (cred, region, bucket_name) = get_config(&url)?;
+    Ok(Box::new(
+        S3Store::new(&url, &bucket_name, region, cred).await?,
+    ))
+}
+
+pub fn make(url: &str) -> Pin<Box<dyn Future<Output = Result<Box<dyn Store>>>>> {
+    Box::pin(make_inner(url.into()))
+}
+
+#[derive(Clone)]
+struct S3Store {
+    bucket: Bucket,
+    url: String,
+}
+
+impl S3Store {
+    pub async fn new(
+        url: &str,
+        bucket_name: &str,
+        region: Region,
+        cred: Credentials,
+    ) -> Result<Self> {
+        let bucket = Bucket::new(bucket_name, region, cred)?.with_path_style();
+
+        Ok(Self {
+            bucket: bucket,
+            url: url.to_owned(),
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Store for S3Store {
+    async fn get(&self, key: &[u8]) -> super::Result<Vec<u8>> {
+        let response = self.bucket.get_object(std::str::from_utf8(key)?).await?;
```

Comment:
Key is not granted to be convertable to utf8 if the get_object REQUIRES the key to be a string, I assume it's better to do a `hex` of the provided binary. doing utf8(key) will most definitely fail when running